cpp/src/arrow/compute/exec.cc

```diff
@@ -156,15 +156,22 @@ Result<ExecBatch> ExecBatch::Make(std::vector<Datum> values) {
 
 Result<std::shared_ptr<RecordBatch>> ExecBatch::ToRecordBatch(
     std::shared_ptr<Schema> schema, MemoryPool* pool) const {
+  if (static_cast<size_t>(schema->num_fields()) > values.size()) {
+    return Status::Invalid("mismatching schema size");
+  }
   ArrayVector columns(schema->num_fields());
 
   for (size_t i = 0; i < columns.size(); ++i) {
     const Datum& value = values[i];
     if (value.is_array()) {
       columns[i] = value.make_array();
       continue;
+    } else if (value.is_scalar()) {
+      ARROW_ASSIGN_OR_RAISE(columns[i],
+                            MakeArrayFromScalar(*value.scalar(), length, pool));
+    } else {
+      DCHECK(false);
```

Comment:
```suggestion
      Unreachable();
```

Reply:
Well, it's not really unreachable :-)

Reply:
It would be a violation of ExecBatch's class invariant for the values to be other than Array or Scalar. Now that I'm looking for a statement of that invariant it's not easy to point at something, the closest I've got is in [streaming_execution.rst](https://github.com/apache/arrow/blob/f941118ea6ffbe1d1d8367d0218566e9e9dae550/docs/source/cpp/streaming_execution.rst#L317-L319). The constructor and ExecBatch::Make don't enforce this either. This validation should be explicit and centralized in ExecBatch

Reply:
We should certainly add `ExecBatch::Validate` (and perhaps `ExecBatch::ValidateFull`).

Reply:
Ok. Let's at least add a meaningful error message to the `DCHECK` :-)

Reply:
Created https://issues.apache.org/jira/browse/ARROW-18015 for the validation. In the meantime, I added a meaningful error message.